phantom-drift/introspection.proto

```diff
@@ -0,0 +1,451 @@
+            Connection conn = 1;
+            // the parent connection by reference.
+            bytes conn_id   = 2;
+        }
+    }
+
+    // Timeline contains the timestamps of the well-known milestones of a stream.
+    message Timeline {
+        // the instant when the stream was opened.
+        google.protobuf.Timestamp open_ts = 1;
+        // the instant when the stream was terminated.
+        google.protobuf.Timestamp close_ts = 2;
+    }
+
+    // the id of this stream, not to be shown in user tooling,
+    // used for (cross)referencing streams.
+    bytes id        = 1;
+    // the protocol pinned to this stream.
+    string protocol = 2;
+    // our role in this stream.
+    Role role       = 3;
+    // traffic statistics.
+    Traffic traffic = 4;
+    // the connection this stream is hosted under.
+    ConnectionRef conn  = 5;
+    // the timeline of the stream, see Stream.Timeline.
+    Timeline timeline   = 6;
+
+    // the instantaneous latency of this stream in nanoseconds.
+    // TODO: this is hard to calculate.
+    uint64 latency_ns       = 16;
+    // user provided tags.
+    repeated string user_provided_tags  = 99;
+}
+
+// DHT metrics and state.
+message DHT {
+
+    message Params {
+        // maximum number of requests to perform.
+        uint64 k                    = 1;
+        // concurrency of asynchronous requests.
+        uint64 alpha                = 2;
+        // number of disjoint paths to use.
+        uint64 disjoint_paths       = 3;
+    }
+
+    message RoutingTable {
+
+        message Bucket {
+            // bucket peers.
+            repeated bytes peers    = 1;
+        }
+
+        // size of the buckets.
+        uint64 bucket_size          = 1;
+        // peers queried.
+        repeated Bucket buckets     = 2;
+    }
+
+    message Operations {
+        // counter for put operations.
+        ResultCounter operation_put = 1;
+        // latency for put operations.
+        repeated uint64 latency_put_ns = 2;
+        // counter for get operations.
+        ResultCounter operation_get = 3;
+        // latency for get operations.
+        repeated uint64 latency_get_ns = 4;
+        // counter for provide operations.
+        ResultCounter operation_provide = 5;
+        // latency for provide operations.
+        repeated uint64 latency_provide_ns = 6;
+        // counter for provide operations.
+        ResultCounter operation_provide = 7;
+        // latency for provide operations.
+        repeated uint64 latency_provide_ns = 8;
+    }
+
+    message Query {
+
+        // Trigger of the query.
+        enum Trigger {
+            API                     = 0;
+            DISCOVERY               = 1;
+        }
+
+        // Type of the query.
+        enum Type {
+            CONTENT                 = 0;
+            PROVIDER                = 1;
+            VALUE                   = 2;
+        }
+
+        // id of the query.
+        bytes id                    = 1;
```

Comment:
ID of the target of the query, right?

Reply:
Since the type is `bytes` and elsewhere peer Ids are `string`, I suspect this is intended for internal referencing, like connection and stream ids? In which case, we should add a `string` field like `target_peer_id`?

(related: should the peer lists in Query and Bucket in the DHT section be `repeated string` not `repeated bytes`?)

Reply:
This ID can represent a peer ID or a CID. I'm fine making these strings for now, so that they're presentable on UI immediately. If we make them bytes, the consumer would be burdened with converting to a string prior to presenting.

Reply:
Note that this object can represent three types of queries: peer, provider, key/value.